internal/pkg/controllers/execLog.go

```diff
@@ -0,0 +1,218 @@
+	statusFailed                   = "failed"
+)
+
+var (
+	ErrOverwriteNotConfirmed = errors.New("write execution log requested but overwrite not confirmed")
+	ErrClientIDsNotConfirmed = errors.New("client IDs from previous execution log not confirmed")
+)
+
+type ExecutionLog struct {
+	params       *options.ParameterBag
+	logFilename  string
+	promptReader config.PromptReader
+	hostInfo     *config.HostInfo
+
+	exists             bool
+	overwriteConfirmed bool
+
+	logInfo *ExecutionLogInfo
+}
+
+type ExecutionLogInfo struct {
+	ExecutedAt time.Time     `yaml:"executed_at"`
+	ExecutedBy string        `yaml:"executed_by"`
+	ExecutedOn string        `yaml:"executed_on"`
+	APIUser    string        `yaml:"api_user"`
+	APIURL     string        `yaml:"api_url"`
+	APIAuth    string        `yaml:"api_auth"`
+	NumClients int           `yaml:"num_clients"`
+	Failed     int           `yaml:"failed"`
+	Jobs       []*models.Job `yaml:"jobs"`
+}
+
+func NewExecLog(params *options.ParameterBag,
+	logFilename string,
+	promptReader config.PromptReader,
+	hostInfo *config.HostInfo) (el *ExecutionLog) {
+	el = &ExecutionLog{
+		params:       params,
+		logFilename:  logFilename,
+		promptReader: promptReader,
+		hostInfo:     hostInfo,
+
+		// assume overwrite unles modified to be otherwise
+		overwriteConfirmed: true,
+
+		logInfo: &ExecutionLogInfo{},
+	}
+
+	el.exists = fs.FileExists(el.logFilename)
+
+	return el
+}
+
+func (el *ExecutionLog) ExistingLog() (exists bool) {
+	return el.exists
+}
+
+func (el *ExecutionLog) ConfirmOverwrite() (confirmed bool, err error) {
+	if el.promptReader == nil || config.ReadNoPrompt(el.params) {
+		return true, nil
+	}
+
+	el.overwriteConfirmed, err = el.promptReader.ReadConfirmation(overwriteExistingLogFileMsg)
+	if err != nil {
+		return false, err
+	}
+	if !el.overwriteConfirmed {
+		return false, ErrOverwriteNotConfirmed
+	}
+
+	return true, nil
+}
+
+func (el *ExecutionLog) ShouldWriteLog() (should bool) {
+	return !el.exists || (el.exists && el.overwriteConfirmed)
+}
+
+func (el *ExecutionLog) getAuthInfo(params *options.ParameterBag) string {
+	if params.ReadString(config.APIToken, "") == "" {
+		return "bearer+jwt"
+	}
+	return "basic+apitoken"
+}
+
+func (el *ExecutionLog) SetJobs(executionResults []*models.Job) {
+	el.logInfo.Jobs = executionResults
+}
+
+func (el *ExecutionLog) ReadFile() (logInfo *ExecutionLogInfo, err error) {
+	fileContents, err := os.ReadFile(el.logFilename)
+	if err != nil {
+		return nil, err
+	}
+
+	logInfo = &ExecutionLogInfo{}
+	err = yaml.Unmarshal(fileContents, &logInfo)
+	if err != nil {
+		return nil, err
+	}
+	el.logInfo = logInfo
+
+	return el.logInfo, nil
+}
+
+func (el *ExecutionLog) WriteExecLog(executedAt time.Time, jobs []*models.Job) (err error) {
+	el.SetJobs(jobs)
+	if len(el.logInfo.Jobs) > 0 {
+		err = el.MakeExecutionInfoHeader(executedAt)
+		if err != nil {
+			return err
+		}
+
+		contents, err := yaml.Marshal(el.logInfo)
```

Comment:
On the header of the log we have snake case. Good. 
Shouldn't we convert the fields on the `jobs` array to snake case too? Improves readability.
E.g. `finished_at` or `created_by`.

Reply:
done. added yaml tags to the job model struct. it will impact the regular yaml output on the console, but i think that's ok and aligns the yaml with the json which is a good thing.